### PR TITLE
Adds ERT workflow for accuracy metric calculation

### LINKED ISSUE #29: FlowNet accuracy metric generated using ERT workflows during AHM
Original author: @wouterjdb

We should include the accuracy metric in ERT using workflows during the AHM.

- run after the simulations in an iteration are finished
- output
    - the metrics ({RSME, MSE, ...} for various targets {WOPR, WWPR, WBHP, TOTAL, ...?}) in a single csv file that is appended to
    - a picture that is updated

We should also considering to include the heatmap and/or other useful analytical plots. @edubarrosTNO, which ones were most useful?

Please have a look at to see how you can add a workflow in the code base.

### PR DESCRIPTION
**Pull request by edubarrosTNO**
Resolves #29.

### CODE REVIEW

**Comment by edubarrosTNO** [comment]
@anders-kiaer , after my latest commit, the code passes the code styling test (where it was previously failing). But now it fails one of the running tests. There seems to be some problem with dependency of `webviz` (wrong version of `dash` package). It is weird, because I did not touch that part of the code. I know that you are familiar with `webviz`: can you please take a look at it? Other than that, the PR should be ready for review

File: examples/brugge_parameters.yml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    min: 6.0e-6
+    max: 5.0e-5
+  aquifer:
+    scheme: individual
+    fraction: 0.25
+    delta_depth: 1000
+    size_in_bulkvolumes:
+      min: 1.0e-6
+      max: 2
+      loguniform: True
```
Comment (anders-kiaer) [inline examples/brugge_parameters.yml:97]: Suggest the Brugge configuration file is added in a separate pull request.

File: src/flownet/ahm/_assisted_history_matching.py [hunk 2]
```diff
@@ -1,18 +1,25 @@
...
-import argparse
+import re
 import shutil
-import concurrent.futures
+import subprocess
 from typing import List, Dict, Optional, Tuple
+import yaml
+import matplotlib.pyplot as plt
+from matplotlib.ticker import ScalarFormatter
+from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
```
Comment (anders-kiaer) [inline src/flownet/ahm/_assisted_history_matching.py:16]: When we first change the order here of imports, maybe we can group them according to PEP8. 
> Imports should be grouped in the following order:
> 
> Standard library imports.
> Related third party imports.
> Local application/library specific imports.
> You should put a blank line between each group of imports.

File: src/flownet/ahm/_assisted_history_matching.py [hunk 3]
```diff
@@ -270,3 +277,450 @@ def save_iteration_parameters():
...
+    Returns:
+        A Pandas dataframe containing only values within the specified range
+        for the provided column
+    """
+
+    df_out = df_in[
+        (df_in[key_filter].values >= min_value) & (df_in[key_filter].values < max_value)
+    ]
+
+    return df_out
```
Comment (anders-kiaer) [inline src/flownet/ahm/_assisted_history_matching.py:303]:
```suggestion
    return df_in[
        (df_in[key_filter].values >= min_value) & (df_in[key_filter].values < max_value)
    ]
```

File: src/flownet/ahm/_assisted_history_matching.py [hunk 4]
```diff
@@ -270,3 +277,450 @@ def save_iteration_parameters():
...
+            score = math.sqrt(score)
+            if metric == "NRMSE":
+                score = score / (np.amax(data_reference) - np.amin(data_reference))
+    elif metric in ("MAE", "NMAE"):
+        score = mean_absolute_error(data_reference, data_test)
+        if metric == "NMAE":
+            score = score / (np.amax(data_reference) - np.amin(data_reference))
+    elif metric == "R2":
+        score = r2_score(data_reference, data_test, multioutput="variance_weighted")
+
```
Comment (anders-kiaer) [inline src/flownet/ahm/_assisted_history_matching.py:441]:
```
else:
    raise ValueError(f"Unknown metric {metric}")
```

File: src/flownet/ahm/_assisted_history_matching.py [hunk 5]
```diff
@@ -270,3 +277,450 @@ def save_iteration_parameters():
...
+    Args:
+        runpath: Path to where the realization is run.
+        ecl_base: Path to where the realization is run.
+    Returns:
+        EclSum
+    """
+
+    eclsum = EclSum(str(pathlib.Path(runpath) / pathlib.Path(ecl_base)))
+
+    return eclsum
```
Comment (anders-kiaer) [inline src/flownet/ahm/_assisted_history_matching.py:457]:
```suggestion
    return EclSum(str(pathlib.Path(runpath) / pathlib.Path(ecl_base)))
```

File: src/flownet/ahm/_assisted_history_matching.py [hunk 6]
```diff
@@ -270,3 +277,450 @@ def save_iteration_parameters():
...
+        )
+        df_metrics = df_metrics.append(dict_metric_tmp, ignore_index=True)
+
+        # Plotting accuracy metrics over iterations
+        save_plots_metrics(df_metrics, metrics, str_key)
+
+    # Saving accuracy metrics to CSV file
+    df_metrics.to_csv(args.outfile + ".csv", index=False)
+
+    print("[Done]")
```
Comment (anders-kiaer) [inline src/flownet/ahm/_assisted_history_matching.py:726]: In order to not let `_assisted_history_matching` grow too big, maybe we should add the accuracy metric calcutation in a separate module?

File: src/flownet/static/SAVE_ITERATION_ANALYTICS_WORKFLOW [hunk 7]
```diff
@@ -0,0 +1 @@
+SAVE_ITERATION_ANALYTICS_WORKFLOW_JOB <RUN_PATH> <ECL_BASE> <YAML_OBS> <ANALYSIS_START> <ANALYSIS_END> <ANALYSIS_QUANTITY> <ANALYSIS_METRIC> <ANALYSIS_OUTFILE>
```
Comment (anders-kiaer) [inline src/flownet/static/SAVE_ITERATION_ANALYTICS_WORKFLOW:1]: Missing empty blank line at end of file.

File: src/flownet/static/SAVE_ITERATION_ANALYTICS_WORKFLOW_JOB [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+MIN_ARG       8
+MAX_ARG       8
+ARG_TYPE      0 STRING
+ARG_TYPE      1 STRING
+ARG_TYPE      2 STRING
+ARG_TYPE      3 STRING
+ARG_TYPE      4 STRING
+ARG_TYPE      5 STRING
+ARG_TYPE      6 STRING
+ARG_TYPE      7 STRING
```
Comment (anders-kiaer) [inline src/flownet/static/SAVE_ITERATION_ANALYTICS_WORKFLOW_JOB:12]: Missing empty blank line at end of file.

**Comment by anders-kiaer** [comment]
Superseded by #44.
